core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -360,41 +359,46 @@ class KafkaApis(val requestChannel: RequestChannel,
             requestChannel.noOperation(request.processor, request)
           }
         } else {
-          val response = ProducerResponse(produceRequest.correlationId,
-                                          mergedResponseStatus,
-                                          produceRequest.versionId,
-                                          delayTimeMs)
-          requestChannel.sendResponse(new RequestChannel.Response(request,
-                                                                  new RequestOrResponseSend(request.connectionId,
-                                                                                            response)))
+          val respHeader = new ResponseHeader(request.header.correlationId)
+          val respBody = if (request.header.apiVersion > 0)
```

Comment:
Indeed, explicitly handling each version is safer. I'll update the logic and log an error/throw an exception when `apiVersion` is unknown.